### PR TITLE
Key Value 2.0

### PR DESCRIPTION
**Pull request by jeffreyssmith2nd**
## Change Description

- Add reverse iterators
- Add a singleton based on KV storage
- Update reflection/serialization code
- Bring in general improvements from staging-c
- Add `non_unique` type

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/eosiolib/core/eosio/to_key.hpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+template <typename T, typename S>
+void to_key_range(const T& obj, datastream<S>& stream) {
+   for (const auto& elem : obj) { to_key_optional(&elem, stream); }
+   to_key_optional((decltype(&*std::begin(obj))) nullptr, stream);
+}
+
+template <typename T, typename S>
+void to_key(const std::vector<T>& obj, datastream<S>& stream) {
+   for (const T& elem : obj) { to_key_optional(&elem, stream); }
```
[the remark this answers is not in the document]
Comment (larryk85) [reply libraries/eosiolib/core/eosio/to_key.hpp:124]: All of these "ranged" types can just use the same overload and a small sfinae helper can be made to filter for these overloads.

Suggested solution:

``` c++
namespace detail {
   template <template <typename> class C, typename T>
   constexpr bool is_ranged_type( C<T> ) { 
      using type = std::decay_t<C><T>;
      return std::is_same_v<std::vector<T>, type> || do the same for the other types.
  }
}

template <typename T, typename S>
auto to_key( const T& obj, datastream<S>& stream ) -> std::enable_if_t<is_ranged_type(std::declval<T>()), void> {
   to_key_range(obj, stream);
}

```

File: libraries/eosiolib/core/eosio/to_key.hpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+template <typename T, typename S>
+void to_key_optional(const bool* obj, datastream<S>& stream) {
+   if (obj == nullptr)
+      stream.write('\0');
+   else if (!*obj)
+      stream.write('\1');
+   else {
+      stream.write("\2", 2);
```
Comment (larryk85) [inline libraries/eosiolib/core/eosio/to_key.hpp:83]: These look like bugs. Will this write {\2, \0}? And is that what we want?

Comment (jeffreyssmith2nd) [reply libraries/eosiolib/core/eosio/to_key.hpp:83]: As discussed on the call, this does seem to be a bug that was pulled in from old code. I've brought over any remaining differences to include fixing this.

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 3]
```diff
@@ -73,281 +74,338 @@ namespace eosio {
...
 
 /**
  * The key_type struct is used to store the binary representation of a key.
  */
-struct key_type : private std::string {
+struct key_type : private std::vector<char> {
    key_type() = default;
-   key_type(const char* c, size_t s) : std::string(c, s) {}
+
+   explicit key_type(std::vector<char>&& v) : std::vector<char>(v) {}
```
[the remark this answers is not in the document]
Comment (jeffreyssmith2nd) [reply libraries/eosiolib/contracts/eosio/key_value.hpp:127]: I don't think the underlying vector constructor is constexpr

```
error: constexpr constructor never produces a constant expression [-Winvalid-constexpr]
note: non-constexpr constructor 'vector' cannot be used in a constant expression
```

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 4]
```diff
@@ -73,281 +74,338 @@ namespace eosio {
...
       return ret;
    }
 
+   key_type& operator+=(const key_type& b) {
+      this->insert(this->end(), b.begin(), b.end());
+      return *this;
+   }
+
    bool operator==(const key_type& b) const {
       return size() != b.size() && memcmp(data(), b.data(), b.size()) == 0;
```
[the remark this answers is not in the document]
Comment (jeffreyssmith2nd) [reply libraries/eosiolib/contracts/eosio/key_value.hpp:143]: Typo, good catch

**Comment by jeffreyssmith2nd** [comment]
😢 This didn't show merge conflicts when I tested the merge locally. Not sure why Github thinks there are conflicts.

**Pull request merged**